<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>GWAS preprocessing for seismicGWAS input • seismicGWAS</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="GWAS preprocessing for seismicGWAS input">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">seismicGWAS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/seismicGWAS.html">Introduction</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-data-processing" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Data processing</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-data-processing">
<li><a class="dropdown-item" href="../articles/GWAS_processing.html">GWAS preprocessing</a></li>
    <li><a class="dropdown-item" href="../articles/scRNA-seq_processing.html">scRNA-seq preprocessing</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ylaboratory/seismic/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>GWAS preprocessing for seismicGWAS input</h1>
                        <h4 data-toc-skip class="author">Qiliang Lai</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/ylaboratory/seismic/blob/HEAD/vignettes/GWAS_processing.Rmd" class="external-link"><code>vignettes/GWAS_processing.Rmd</code></a></small>
      <div class="d-none name"><code>GWAS_processing.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <em>seismic</em> framework takes the output from <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004219" class="external-link">MAGMA</a> gene analysis software as an input.<br>
Basically, the MAGMA software digests SNP-level summary statistics of a trait in a genome-wide association study into a gene-level MAGMA z-score statistics. This tutorial will guide you through the preprocessing steps necessary to convert raw GWAS summary statistics into the proper format for <em>seismic</em> analysis. Before beginning, ensure you have installed the MAGMA software and obtained the required files: - Gene location file - SNP synonyms file (dbSNP) - Population-specific reference data These files can be all downloaded from the <a href="https://cncr.nl/research/magma/" class="external-link">MAGMA website</a>.</p>
</div>
<div class="section level2">
<h2 id="step-1-clean-up-the-gwas-summary-statistics-file">Step 1: Clean up the GWAS summary statistics file<a class="anchor" aria-label="anchor" href="#step-1-clean-up-the-gwas-summary-statistics-file"></a>
</h2>
<p>The minimum required information in your GWAS summary statistics file includes: - Chromosome ID column - Base-pair position column - P value column Additionally, we strongly recommend including SNP IDs (rsIDs) to facilitate MAGMA analysis. Here’s an example of how your data might look:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#load packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span> <span class="co">#for lift over</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#load data</span></span>
<span><span class="va">example_gwas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"inst/articles/data/sample_gwas.txt"</span>, header <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>  <span class="co">#reaplce it with your local path</span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> </span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">example_gwas</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 7</span></span></span>
<span><span class="co">##   Allele1 Allele2 Direction P.value  group Chromosome     BP</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> t       c       ---        0.054<span style="text-decoration: underline;">1</span> <span style="text-decoration: underline;">184</span>129          1 <span style="text-decoration: underline;">796</span>338</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> a       g       +++        0.031<span style="text-decoration: underline;">5</span>  <span style="text-decoration: underline;">46</span>089          1 <span style="text-decoration: underline;">817</span>341</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> t       c       +++        0.015<span style="text-decoration: underline;">9</span> <span style="text-decoration: underline;">189</span>219          1 <span style="text-decoration: underline;">817</span>514</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> a       t       -+-        0.199   <span style="text-decoration: underline;">39</span>112          1 <span style="text-decoration: underline;">870</span>176</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> c       g       --+        0.730   <span style="text-decoration: underline;">80</span>434          1 <span style="text-decoration: underline;">928</span>622</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> t       c       --+        0.913  <span style="text-decoration: underline;">185</span>512          1 <span style="text-decoration: underline;">996</span>120</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="step-2-map-the-snp-location-to-grch37">Step 2: Map the SNP location to GRCh37<a class="anchor" aria-label="anchor" href="#step-2-map-the-snp-location-to-grch37"></a>
</h2>
<p>For MAGMA gene analysis, the required population-specific reference data is based on the GRCh37 genome build, all SNP location should be converted to GRCh37 (Hg19). The example data file shown above is a GWAS summary statistics file based on the GRCh38 genome build. To map the SNP location to GRCh37, we can use the <a href="https://genome.ucsc.edu/cgi-bin/hgLiftOver" class="external-link">LiftOver tool</a> provided by UCSC. Here is an example of how to convert the SNP location with the chain object “hg38ToHg19.over.chain.gz” downloaded from the <a href="https://hgdownload.cse.ucsc.edu/goldenpath/hg19/liftOver/" class="external-link">UCSC website</a>, together with the rtrackler package in R.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#load the chain object</span></span>
<span><span class="va">chainObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/Chain-class.html" class="external-link">import.chain</a></span><span class="op">(</span><span class="st">"inst/articles/data/hg38ToHg19.over.chain"</span><span class="op">)</span> <span class="co">#your downloaded chain object</span></span>
<span></span>
<span><span class="co">#Granges object with the bed information. The seqnames are named with "chr" prefix.</span></span>
<span><span class="va">grObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span>seqnames<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="va">example_gwas</span><span class="op">$</span><span class="va">Chromosome</span><span class="op">)</span>,</span>
<span>                   ranges<span class="op">=</span><span class="fu">IRanges</span><span class="op">(</span>start<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">example_gwas</span><span class="op">$</span><span class="va">BP</span><span class="op">)</span>, end<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">example_gwas</span><span class="op">$</span><span class="va">BP</span><span class="op">)</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#lift over</span></span>
<span><span class="va">hg19_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/liftOver.html" class="external-link">liftOver</a></span><span class="op">(</span><span class="va">grObject</span>, <span class="va">chainObject</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#extract the result data frame, select the necessary columns</span></span>
<span><span class="va">hg19_res_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">hg19_res</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="co">#get the result data frame</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="va">seqnames</span>, pattern <span class="op">=</span> <span class="st">"chr"</span>,replacement <span class="op">=</span> <span class="st">""</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="co">#change pattern back</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">group</span>, <span class="va">start</span>, <span class="va">seqnames</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span> <span class="st">"start"</span> <span class="op">=</span> <span class="st">"BP"</span> , <span class="st">"seqnames"</span> <span class="op">=</span> <span class="st">"Chromosome"</span> <span class="op">)</span> <span class="co">#rename columns</span></span>
<span></span>
<span><span class="co">#get the 1 to 1 mapping</span></span>
<span><span class="va">hg19_res_df</span> <span class="op">&lt;-</span> <span class="va">hg19_res_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">group</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">==</span><span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="op">)</span> </span>
<span></span>
<span><span class="co">#merge with the original data frame (the group column indicates the original position in the table)</span></span>
<span><span class="va">example_gwas</span> <span class="op">&lt;-</span> <span class="va">example_gwas</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>group <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>  <span class="co">#original order </span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">Chromosome</span>, <span class="op">-</span><span class="va">BP</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="co">#remove the original column</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">hg19_res_df</span> , by <span class="op">=</span> <span class="st">"group"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="co">#left join annotation</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">BP</span><span class="op">)</span><span class="op">)</span> <span class="co">#filter unmapped data</span></span>
<span></span>
<span><span class="co">#write out for later</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">example_gwas</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>BP <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">BP</span><span class="op">)</span><span class="op">)</span>,  </span>
<span>            file <span class="op">=</span> <span class="st">"inst/articles/data/sample_gwas_hg19.txt"</span>,  <span class="co">#replace it with your own path</span></span>
<span>            sep<span class="op">=</span><span class="st">"\t"</span>, col.names <span class="op">=</span> <span class="cn">T</span>, row.names <span class="op">=</span> <span class="cn">F</span>, quote <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">example_gwas</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 7</span></span></span>
<span><span class="co">##   Allele1 Allele2 Direction P.value group     BP Chromosome</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> t       c       ---        0.054<span style="text-decoration: underline;">1</span>     1 <span style="text-decoration: underline;">731</span>718 1         </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> a       g       +++        0.031<span style="text-decoration: underline;">5</span>     2 <span style="text-decoration: underline;">752</span>721 1         </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> t       c       +++        0.015<span style="text-decoration: underline;">9</span>     3 <span style="text-decoration: underline;">752</span>894 1         </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> a       t       -+-        0.199      4 <span style="text-decoration: underline;">805</span>556 1         </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> c       g       --+        0.730      5 <span style="text-decoration: underline;">864</span>002 1         </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> t       c       --+        0.913      6 <span style="text-decoration: underline;">931</span>500 1</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="step-3-map-the-snp-location-to-snp-id">Step 3: Map the SNP location to SNP ID<a class="anchor" aria-label="anchor" href="#step-3-map-the-snp-location-to-snp-id"></a>
</h2>
<p>If the SNP ID is not included in the GWAS summary statistics file, we can use the SNP synonyms file to map the SNP location to SNP ID. To simplify the procedure, we implement a bash script in <code>tools/annotate_SNP.sh</code> for the step, where <a href="https://bedtools.readthedocs.io/en/latest/" class="external-link">bedtools</a> is required. This script will require the dbSNP file and the GWAS summary statistics file as input, and output a new file with SNP ID included in the last column. You can find this script in the <a href="https://github.com/ylaboratory/seismic-analysis" class="external-link">seismic-analysis github repo</a>, under the <code>src/tools</code> directory. To view all specific parameters of the script, you can run the following command in the terminal:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">bash</span> inst/articles/code/annotate_SNP.sh <span class="at">-h</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##     This script help you locate SNP and transform the locus (by chromosome id and location) to the corresponding variant rsid.</span></span>
<span><span class="co">##     Only intersected SNPs will be printed out.</span></span>
<span><span class="co">##     Note!!!! 1. All file path should be the same as the original files. 2. bedtools is required, so make sure it's in your environment or path</span></span>
<span><span class="co">##     Here are the required parameters:</span></span>
<span><span class="co">##     -g | GWAS summary statistics files. Or any other files that you would like to annotate SNPs to.</span></span>
<span><span class="co">##     -o | Outpur file header. If the value is given as 'dir1/dir2/output', then the final output file will be 'dir1/dir2/output.annot.table.txt'. If it is not specified, the output file will be the same as the input file with the suffix ignored.</span></span>
<span><span class="co">##     -c | Chromosome column, i.e. the column with with chromosome information be (in the table). The value should be specified or by default it will be 1.</span></span>
<span><span class="co">##     -p | Position column, i.e. The column with the the position information (in the table), The value should be specified or by default it will be 2.</span></span>
<span><span class="co">##     -b | Basis of chromosome numbering, to specify the if first position on a chromosome is labeled as '1'(1-base), or '0'(0-base). The default parameters are start with 1. #only these two are available!</span></span>
<span><span class="co">##     -k | Keep all unmapped SNPs which do not exist in the dbSNP file or not. By default it is True. (You can choose between 'True' or 'False')</span></span>
<span><span class="co">##     </span></span>
<span><span class="co">##     optional parameters:</span></span>
<span><span class="co">##     -v | VCF file for the dbsnp version you used. For default it will be located in 'data/ref/Hg19.dbsnp151.vcf'. This can be downloaded online.</span></span>
<span><span class="co">##     -n | Number of tasks to split the job. The default value is 1. If the memory is not enough, you can split the job into several parts but it will take more time.</span></span>
<span><span class="co">##     -t | Customized bedtools command. By default it is 'bedtools'. If you do not have it in your environment, you can specify the path to the bedtools command.</span></span>
<span><span class="co">##     -w | The working directory. By default it is the current directory. If you want to specify the working directory, you can use this parameter.</span></span>
<span><span class="co">##     -h | Help message.</span></span>
<span><span class="co">## </span></span></code></pre>
<p>Here is an example of how to run the script with the example data file:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">bash</span> inst/articles/code/annotate_SNP.sh <span class="at">-g</span> inst/articles/data/sample_gwas_hg19.txt <span class="at">-o</span> inst/articles/data/sample_gwas_hg19 <span class="dt">\</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>-c 6 <span class="at">-p</span> 7 <span class="at">-v</span> data/ref/Hg19.dbsnp151.vcf <span class="at">-t</span> bin/bedtools <span class="co">#replace it with your own path</span></span></code></pre></div>
<p>Note that the script may takes hours to run and consumes more than 100 GB of memory. To alleviate memory usage (but increase the running time), the <code>-n</code> option may be used to sequentially split the task into several. The annotated table will be printed with the path and header specified by the <code>-o</code> option.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> inst/articles/data/sample_gwas_hg19.annot.table <span class="co">#replace it with your own path</span></span></code></pre></div>
<pre><code><span><span class="co">## Allele1  Allele2 Direction   P.value group   Chromosome  BP  marker.name</span></span>
<span><span class="co">## t    c   --- 0.0541  1   1   731718 rs58276399</span></span>
<span><span class="co">## a    g   +++ 0.03149 2   1   752721 rs3131972</span></span>
<span><span class="co">## t    c   +++ 0.01587 3   1   752894 rs3131971</span></span>
<span><span class="co">## a    t   -+- 0.1987  4   1   805556 rs72631880</span></span>
<span><span class="co">## c    g   --+ 0.7303  5   1   864002 rs1806501</span></span>
<span><span class="co">## t    c   --+ 0.913   6   1   931500 rs74718486</span></span>
<span><span class="co">## a    g   --+ 0.4902  7   1   938125 rs2710868</span></span>
<span><span class="co">## a    g   --+ 0.785   8   1   941334 rs57683598</span></span>
<span><span class="co">## t    c   +++ 0.1863  9   1   948921 rs15842</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="step-4-generate-the-input-file-for-magma">Step 4: Generate the input file for MAGMA<a class="anchor" aria-label="anchor" href="#step-4-generate-the-input-file-for-magma"></a>
</h2>
<p>The software takes two input files generated from the previously processed (or unprocessed but with the required columns) GWAS summary statistics file: a SNP location file and a P value file. You can simply use the <code>awk</code> command to extract the information, for example:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co">#print rsID, chromosome, BP but not colum names:</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="co">#starting from the second line (NR&gt;1), field 8, 6, 7</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="fu">awk</span> <span class="st">'{if (NR&gt;1) print $8, $6, $7}'</span> inst/articles/data/sample_gwas_hg19.annot.table <span class="op">&gt;</span> inst/articles/data/sample_gwas_hg19.snp_loc.txt <span class="co">#replace it with your own path</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="co">#print rsID, Pvalue but not colum names:</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="co">#starting from the second line (NR&gt;1), field 8, 4</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="fu">awk</span> <span class="st">'{if (NR&gt;1) print $8, $4}'</span> inst/articles/data/sample_gwas_hg19.annot.table <span class="op">&gt;</span>inst/articles/data/sample_gwas_hg19.p_val.txt <span class="co">#replace it with your own path</span></span></code></pre></div>
<p>or in R:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#replace it with your own path</span></span>
<span><span class="va">anno_tbl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"inst/articles/data/sample_gwas_hg19.annot.table"</span>, header<span class="op">=</span><span class="cn">T</span><span class="op">)</span> </span>
<span><span class="va">anno_tbl</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">marker.name</span>, <span class="va">Chromosome</span>, <span class="va">BP</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"inst/articles/data/sample_gwas_hg19.snp_loc.txt"</span>, col.names <span class="op">=</span> <span class="cn">F</span>, row.names <span class="op">=</span> <span class="cn">F</span>, quote <span class="op">=</span> <span class="cn">F</span>, sep<span class="op">=</span><span class="st">"\t"</span><span class="op">)</span></span>
<span><span class="va">anno_tbl</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">marker.name</span>, <span class="va">P.value</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"inst/articles/data/sample_gwas_hg19.p_val.txt"</span>, col.names <span class="op">=</span> <span class="cn">F</span>, row.names <span class="op">=</span> <span class="cn">F</span>, quote <span class="op">=</span> <span class="cn">F</span>, sep<span class="op">=</span><span class="st">"\t"</span><span class="op">)</span></span></code></pre></div>
<p>Data preview:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>  inst/articles/data/sample_gwas_hg19.snp_loc.txt</span></code></pre></div>
<pre><code><span><span class="co">## rs58276399   1   731718</span></span>
<span><span class="co">## rs3131972    1   752721</span></span>
<span><span class="co">## rs3131971    1   752894</span></span>
<span><span class="co">## rs72631880   1   805556</span></span>
<span><span class="co">## rs1806501    1   864002</span></span>
<span><span class="co">## rs74718486   1   931500</span></span>
<span><span class="co">## rs2710868    1   938125</span></span>
<span><span class="co">## rs57683598   1   941334</span></span>
<span><span class="co">## rs15842  1   948921</span></span>
<span><span class="co">## rs9442364    1   970215</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="step-5-run-magma-software">Step 5: Run MAGMA software<a class="anchor" aria-label="anchor" href="#step-5-run-magma-software"></a>
</h2>
<p>There are two steps: gene annotation and gene analysis. You can use our implemented bash script <code>tools/magma_gene_zscore_analysis.sh</code> to run the two steps together. Another paramter to specificy is the gene analysis window size, which is set as 35 kilobase (kb) upstream to 10 kb downstream by default, as recommend by <a href="https://www.nature.com/articles/s41588-020-0610-9" class="external-link">the previous study</a>. Although we have shown that output files from various window sizes make only a minor difference in the results of <em>seismic</em> analysis, you can still change the window size by specifying the parameter <code>--window-size</code> in the script. To view all specific parameters of the script, you can run the following command in the terminal:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">bash</span> inst/articles/code/magma_gene_zscore_analysis.sh <span class="at">-h</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Usage: inst/articles/code/magma_gene_zscore_analysis.sh -l SNP_LOC_FILE -p SNP_P_VALUE_FILE -o OUTPUT_DIR -s COHORT_SIZE -m MAGMA_PATH -g GENE_LOC_FILE -b B_FILE [-w WINDOW_SIZES] [-h]</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## This script performs MAGMA annotation and analysis on SNP data.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Required parameters:</span></span>
<span><span class="co">##     -l | SNP location file</span></span>
<span><span class="co">##     -p | SNP p-value file</span></span>
<span><span class="co">##     -o | Output directory</span></span>
<span><span class="co">##     -s | Cohort size (use 'cX' format to specify column X containing sample size)</span></span>
<span><span class="co">##     -m | Path to MAGMA binary</span></span>
<span><span class="co">##     -g | Path to gene location file</span></span>
<span><span class="co">##     -b | Path to g1000_eur files (without extension)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Optional parameters:</span></span>
<span><span class="co">##     -w | Window sizes (default: 35,10). Use colon-separated list for multiple sizes, e.g., '35,10:15,20:40,10'</span></span>
<span><span class="co">##     -h | Display this help message</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">bash</span> inst/articles/code/magma_gene_zscore_analysis.sh <span class="at">-l</span> inst/articles/data/sample_gwas_hg19.snp_loc.txt <span class="dt">\</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>-p inst/articles/data/sample_gwas_hg19.p_val.txt <span class="dt">\</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>-o inst/articles/data <span class="dt">\</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>-s 100000 <span class="at">-m</span> bin/magma/magma <span class="at">-g</span> data/ref/magma/NCBI37.3/NCBI37.3.gene.loc <span class="at">-b</span> data/ref/magma/g1000_eur/g1000_eur </span></code></pre></div>
<p>Besides specifying these reference file paths and the curated files from the summary statistics, the total size of the cohort should also be specified. Note that sometimes the cohort size for each SNP may be different, for example, due to some quality control procedure. In that case, a separate column in the p-value file should be added and the input <code>-s</code> argument should tell the script the order of the column (for example, the third column) describing the cohort size. By specifying these arguments, the gene-level MAGMA z-score will be printed out in the output directory with header similar to the snp location file. Here the output file is <code>data/gwas/sample_gwas/sample_gwas_hg19.35.10.genes.out</code>. In addition, please do remember to check the MAGMA running log. Sometimes when the data are not in the correct format, there will be very few genes and SNPs annotated (for example, less than 10% of the SNPs).</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Qiliang Lai, Ruth Dannenfelser, Vicky Yao.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
